File: frontend/src/services/whatsapp.ts
import { fetchApi } from '@/services/api';

export interface Conversation {
  id: string;
  client_id: string;
  client_first_name?: string;
  client_last_name?: string;
  client_phone: string;
  status?: string;
  last_message_at?: string;
  unread_count?: number;
  assigned_user_id?: string;
  last_message_content?: string;
  last_message_direction?: string;
}

export interface Message {
  id: string;
  tenant_id: string;
  conversation_id: string;
  sender_type: string;
  content?: string;
  media_url?: string;
  media_type?: string;
  is_read?: boolean;
  created_at?: string;
  external_id?: string;
  direction?: string;
  status?: string;
  is_ai_generated?: boolean;
}

export interface PaginatedResponse<T> {
  data: T[];
  total: number;
  limit: number;
  offset: number;
}

export const whatsappService = {
  getConversations: async (page = 1, limit = 50): Promise<PaginatedResponse<Conversation>> => {
    return fetchApi(`/whatsapp/conversations?page=${page}&limit=${limit}`);
  },

  getMessages: async (conversationId: string, page = 1, limit = 50): Promise<PaginatedResponse<Message>> => {
    return fetchApi(`/whatsapp/conversations/${conversationId}/messages?page=${page}&limit=${limit}`);
  },

  sendMessage: async (conversationId: string, content: string): Promise<Message> => {
    return fetchApi(`/whatsapp/conversations/${conversationId}/messages`, {
      method: 'POST',
      body: JSON.stringify({ content }),
    });
  },

  getInstanceStatus: async () => {
    return fetchApi('/whatsapp/instance', {
      method: 'GET',
    });
  },

  createInstance: async (instanceName: string) => {
    return fetchApi('/whatsapp/instance', {
      method: 'POST',
      body: JSON.stringify({ instance_name: instanceName }),
    });
  },

  getQr: async () => {
    return fetchApi('/whatsapp/instance/qr', {
      method: 'GET',
    });
  },

  logoutInstance: async () => {
    return fetchApi('/whatsapp/instance/logout', {
      method: 'POST',
    });
  }
};

File: backend/src/models/whatsapp.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Conversation {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub client_id: Uuid,
    pub status: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
    pub last_message_at: Option<DateTime<Utc>>,
    pub unread_count: Option<i32>,
    pub assigned_user_id: Option<Uuid>,
    pub assigned_at: Option<DateTime<Utc>>,
    pub last_agent_response_at: Option<DateTime<Utc>>,
    pub deleted_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Message {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub conversation_id: Uuid,
    pub sender_type: String,
    pub content: Option<String>,
    pub media_url: Option<String>,
    pub media_type: Option<String>,
    pub is_read: Option<bool>,
    pub created_at: Option<DateTime<Utc>>,
    pub external_id: Option<String>,
    pub direction: Option<String>,
    pub status: Option<String>,
    pub is_ai_generated: Option<bool>,
    pub deleted_at: Option<DateTime<Utc>>,
}

// Joined struct for Conversation List endpoint
#[derive(Debug, Serialize, Deserialize)]
pub struct ConversationListItem {
    pub id: Uuid,
    pub client_id: Uuid,
    pub client_first_name: Option<String>,
    pub client_last_name: Option<String>,
    pub client_phone: String,
    pub status: Option<String>,
    pub last_message_at: Option<DateTime<Utc>>,
    pub unread_count: Option<i32>,
    pub assigned_user_id: Option<Uuid>,
    pub assigned_at: Option<DateTime<Utc>>,
    pub last_agent_response_at: Option<DateTime<Utc>>,
    pub last_message_content: Option<String>,
    pub last_message_direction: Option<String>,
}
